dlls/kernelx/kernelx.cpp

```diff
@@ -337,17 +337,121 @@ LPVOID VirtualAllocEx_X(
+    default: result += "UNKNOWN_BASE_PROTECTION"; break;
+    }
+
+    // Extra modifiers (higher bits)
+    if (protect & PAGE_GUARD)
+        result += " | PAGE_GUARD";
+    if (protect & PAGE_NOCACHE)
+        result += " | PAGE_NOCACHE";
+    if (protect & PAGE_WRITECOMBINE)
+        result += " | PAGE_WRITECOMBINE";
+
+    return result;
+}
 
 LPVOID VirtualAlloc_X(
     LPVOID lpAddress,
     SIZE_T dwSize,
-    DWORD  flAllocationType,
-    DWORD  flProtect
+    DWORD flAllocationType,
+    DWORD flProtect
 )
 {
-    return VirtualAllocEx_X(GetCurrentProcess(), lpAddress, dwSize, flAllocationType, flProtect);
+    printf("[VirtualAlloc_X] Request:\n");
+    printf("    - Address Hint : 0x%p\n", lpAddress);
+    printf("    - Size         : %llu bytes (%.2f MB", dwSize, dwSize / (1024.0 * 1024.0));
+    if (dwSize >= (1024ULL * 1024ULL * 1024ULL))
+        printf(" / %.2f GB", dwSize / (1024.0 * 1024.0 * 1024.0));
+    printf(")\n");
+    printf("    - Allocation   : 0x%08X (%s)\n", flAllocationType, std::string(AllocationTypeToString(flAllocationType)).c_str());
+    printf("    - Protection   : 0x%08X (%s)\n", flProtect, ProtectionToString(flProtect).c_str());
+
+    if (dwSize == 0) {
+        printf("[ERROR] Size cannot be zero\n");
+        return nullptr;
+    }
+
+    constexpr DWORD ALLOWED_ALLOCATION_FLAGS = MEM_COMMIT | MEM_RESERVE | MEM_DECOMMIT | MEM_RELEASE |
+        MEM_RESET | MEM_TOP_DOWN | MEM_PHYSICAL | MEM_LARGE_PAGES |
+        MEM_WRITE_WATCH | MEM_RESET_UNDO;
+    constexpr DWORD ALLOWED_PROTECTION_FLAGS = 0xFF | PAGE_GUARD | PAGE_NOCACHE | PAGE_WRITECOMBINE;
+
+    DWORD sanitizedAllocType = flAllocationType & ALLOWED_ALLOCATION_FLAGS;
+    DWORD sanitizedProtect = flProtect & ALLOWED_PROTECTION_FLAGS;
+
+    if (sanitizedAllocType != flAllocationType) {
+        printf("[WARN] Allocation flags sanitized: 0x%08X -> 0x%08X\n", flAllocationType, sanitizedAllocType);
+    }
+    if (sanitizedProtect != flProtect) {
+        printf("[WARN] Protection flags sanitized: 0x%08X -> 0x%08X\n", flProtect, sanitizedProtect);
+    }
+
+    LPVOID result = VirtualAllocEx(GetCurrentProcess(), lpAddress, dwSize, sanitizedAllocType, sanitizedProtect);
+
+    if (!result)
+    {
+        DWORD err = GetLastError();
+        printf("[WARN] VirtualAllocEx failed (%lu), attempting forced fallback...\n", err);
+        
+        DWORD safeFlags = MEM_RESERVE | MEM_COMMIT;
+        DWORD safeProtect = PAGE_READWRITE;
+        result = VirtualAllocEx(GetCurrentProcess(), lpAddress, dwSize, safeFlags, safeProtect);
+
+        if (result)
+        {
+            printf("[SUCCESS] Forced fallback succeeded: Reserved+Committed with PAGE_READWRITE at 0x%p\n", result);
+        }
+        else
+        {
+            printf("[ERROR] Forced fallback also failed: %lu\n", GetLastError());
+        }
+    }
+    else
+    {
+        printf("[SUCCESS] Allocated at: 0x%p\n", result);

```

Comment:
I would not recommend logging successful allocations, it's just noisy and will murder performance in allocation-heavy code.